Core/clim-basic/ports.lisp

```diff
@@ -84,28 +87,20 @@
    (pressed-sheet :initform nil :accessor port-pressed-sheet
 		  :documentation "The sheet the pointer is pressed on, if any")))
 
-(defmethod port-keyboard-input-focus (port)
-  (when (null *application-frame*)
-    (error "~S called with null ~S" 
-           'port-keyboard-input-focus '*application-frame*))
-  (port-frame-keyboard-input-focus port *application-frame*))
-
-(defmethod (setf port-keyboard-input-focus) (focus port)
-  (when (null *application-frame*)
-    (error "~S called with null ~S" 
-           '(setf port-keyboard-input-focus) '*application-frame*))
-  ;; XXX: pane frame is not defined for all streams (for instance not for
-  ;; CLIM:STANDARD-EXTENDED-INPUT-STREAM), so this sanity check would lead to
-  ;; error on that.
-  ;; XXX: also should we allow reading objects from foreign application frames?
-  ;; This was the case on Genera and is requested by users from time to time...
-  #+ (or)
-  (unless (eq *application-frame* (pane-frame focus))
-    (error "frame mismatch in ~S" '(setf port-keyboard-input-focus)))
-  (setf (port-frame-keyboard-input-focus port *application-frame*) focus))
-
-(defgeneric port-frame-keyboard-input-focus (port frame))
-(defgeneric (setf port-frame-keyboard-input-focus) (focus port frame))
+(defgeneric note-input-focus-changed (sheet state)
+  (:documentation "Called when a sheet receives or loses the keyboard
+input focus. This is a McCLIM extension.")
+  (:method (sheet state)
+    (declare (ignore sheet state))))
+
+(defmethod (setf port-keyboard-input-focus) :around (focus port)
+  (let ((old-focus (port-keyboard-input-focus port))
+        (result (call-next-method))     ; clim-null cheats
```

Comment:
Can `clim-null` stop cheating by either returning `nil` or actually storing the focused sheet?

Reply:
it could, but null port has around method for distribute-event which ignores all events. I'd rather not change this behavior in this particular pull request. That means in particular that setf'ing port-keyboard-input-focus does not make any sheet focused, hence port-keyboard-input-focus justifully returns NIL (and part of setf protocol is to return assigned variable). So this "cheat" is arguably the right thing to do. Should I remove the ocmment?